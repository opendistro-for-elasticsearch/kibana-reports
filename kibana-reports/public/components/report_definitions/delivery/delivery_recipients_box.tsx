/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiComboBox } from '@elastic/eui';
import { useEffect } from 'react';

const KibanaRecipientsBox = (props) => {

  const {reportDefinitionRequest} = props;

  const options = [];
  const [selectedOptions, setSelected] = useState([]);

  const onChangeDeliveryRecipients = (selectedOptions) => {
    setSelected(selectedOptions);
  };

  const onCreateDeliveryRecipientOption = (
    searchValue,
    flattenedOptions = []
  ) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
    };

    // Create the option if it doesn't exist.
    if (
      flattenedOptions.findIndex(
        (option) => option.label.trim().toLowerCase() === normalizedSearchValue
      ) === -1
    ) {
      options.push(newOption);
    }

    // Select the option.
    setSelected([...selectedOptions, newOption]);
  };

  useEffect(() => {
    reportDefinitionRequest.delivery.delivery_params = {
      kibana_recipients: selectedOptions
    }
  }, [selectedOptions])

  return (
    <EuiComboBox
      placeholder="Select or create options"
      options={options}
      selectedOptions={selectedOptions}
      onChange={onChangeDeliveryRecipients}
      onCreateOption={onCreateDeliveryRecipientOption}
      isClearable={true}
      data-test-subj="demoComboBox"
    />
  );
};

export { KibanaRecipientsBox };
