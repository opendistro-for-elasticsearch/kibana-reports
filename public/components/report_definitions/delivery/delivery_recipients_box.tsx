/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiComboBox } from '@elastic/eui';

const DeliveryRecipientsBox = () => {
  const options = [];
  const [selectedOptions, setSelected] = useState([]);

  const onChangeDeliveryRecipients = (selectedOptions) => {
    setSelected(selectedOptions);
  };

  const onCreateDeliveryRecipientOption = (
    searchValue,
    flattenedOptions = []
  ) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
    };

    // Create the option if it doesn't exist.
    if (
      flattenedOptions.findIndex(
        (option) => option.label.trim().toLowerCase() === normalizedSearchValue
      ) === -1
    ) {
      options.push(newOption);
    }

    // Select the option.
    setSelected([...selectedOptions, newOption]);
  };

  return (
    <EuiComboBox
      placeholder="Select or create options"
      options={options}
      selectedOptions={selectedOptions}
      onChange={onChangeDeliveryRecipients}
      onCreateOption={onCreateDeliveryRecipientOption}
      isClearable={true}
      data-test-subj="demoComboBox"
    />
  );
};

export { DeliveryRecipientsBox };
