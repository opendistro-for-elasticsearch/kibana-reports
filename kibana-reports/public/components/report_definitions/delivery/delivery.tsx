/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import {
  EuiFieldText,
  EuiFormRow,
  EuiPageHeader,
  EuiTitle,
  EuiPageContent,
  EuiPageContentBody,
  EuiHorizontalRule,
  EuiTextArea,
  EuiCheckbox,
  EuiSpacer,
  EuiComboBox,
  EuiText,
  EuiLink,
  EuiPopover,
  EuiListGroup,
} from '@elastic/eui';
import { DeliveryRecipientsBox } from './delivery_recipients_box';
import { EMAIL_RECIPIENT_OPTIONS } from './delivery_constants';

const INSERT_PLACEHOLDER_OPTIONS = [
  {
    label: 'Report details URL',
    href: '#',
    iconType: 'link',
    size: 's',
  },
  {
    label: 'Report source URL',
    href: '#',
    iconType: 'link',
    size: 's',
  },
  {
    label: 'File download URL',
    href: '#',
    iconType: 'link',
    size: 's',
  },
  {
    label: 'Report creation timestamp',
    href: '#',
    iconType: 'clock',
    size: 's',
  },
];

let delivery_params = {
  subject: '',
  body: '',
  has_attachment: false,
  recipients: ['array'],
};

export function ReportDelivery(props) {
  const { createReportDefinitionRequest } = props;
  const [emailCheckbox, setEmailCheckbox] = useState(false);
  const [emailRecipients, setEmailRecipients] = useState([]);
  const [includeReportAsAttachment, setIncludeReportAsAttachment] = useState(
    false
  );
  const [insertPlaceholder, setInsertPlaceholder] = useState(false);

  // TODO: make these fields not required for Create report definition request so no need for filler values
  let delivery = {
    channel: 'Kibana User',
    delivery_params: {
      subject: 'test subject',
      body: 'test body',
      has_attachment: true,
      recipients: ['test_entry@test.com'],
    },
  };

  const handleEmailCheckbox = (e: {
    target: { checked: React.SetStateAction<boolean> };
  }) => {
    setEmailCheckbox(e.target.checked);
    if (e.target.checked) {
      delivery['channel'] = 'Email';
    }
    delivery['delivery_params'] = delivery_params;
  };

  const handleEmailRecipients = (e: React.SetStateAction<any[]>) => {
    setEmailRecipients(e);
    delivery_params['recipients'].push(e.toString());
  };

  const handleIncludeReportAsAttachment = (e: {
    target: { checked: React.SetStateAction<boolean> };
  }) => {
    setIncludeReportAsAttachment(e.target.checked);
    if (e.target.checked) {
      delivery_params['has_attachment'] = includeReportAsAttachment;
    }
  };

  const handleInsertPlaceholderClick = () => {
    setInsertPlaceholder((insertPlaceholder) => !insertPlaceholder);
  };
  const closeInsertPlaceholder = () => setInsertPlaceholder(false);

  const handleCreateEmailRecipient = (
    searchValue: string,
    flattenedOptions = []
  ) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
    };

    // Create the option if it doesn't exist.
    if (
      flattenedOptions.findIndex(
        (option) => option.label.trim().toLowerCase() === normalizedSearchValue
      ) === -1
    ) {
      EMAIL_RECIPIENT_OPTIONS.push(newOption);
    }

    // Select the option.
    setEmailRecipients([...emailRecipients, newOption]);
  };

  const placeholderInsert = (
    <EuiText size="xs">
      <EuiLink onClick={handleInsertPlaceholderClick}>
        Insert placeholder
      </EuiLink>
    </EuiText>
  );

  const InsertPlaceholderPopover = () => {
    return (
      <div>
        <EuiPopover
          button={placeholderInsert}
          isOpen={insertPlaceholder}
          closePopover={closeInsertPlaceholder}
        >
          <EuiListGroup listItems={INSERT_PLACEHOLDER_OPTIONS} />
        </EuiPopover>
      </div>
    );
  };

  const EmailDelivery = () => {
    const [emailSubject, setEmailSubject] = useState('');
    const [emailBody, setEmailBody] = useState('');

    const handleEmailSubject = (e: {
      target: { value: React.SetStateAction<string> };
    }) => {
      setEmailSubject(e.target.value);
      delivery_params['subject'] = e.target.value.toString();
    };

    const handleEmailBody = (e: {
      target: { value: React.SetStateAction<string> };
    }) => {
      setEmailBody(e.target.value);
      delivery_params['body'] = e.target.value.toString();
    };

    return (
      <div>
        <EuiFormRow
          label="Email recipients"
          helpText="Select or add recipients"
        >
          <EuiComboBox
            placeholder={'Add users here'}
            options={EMAIL_RECIPIENT_OPTIONS}
            selectedOptions={emailRecipients}
            onChange={handleEmailRecipients}
            onCreateOption={handleCreateEmailRecipient}
          />
        </EuiFormRow>
        <EuiSpacer />
        <EuiFormRow label="Email subject">
          <EuiFieldText
            placeholder="Subject line"
            value={emailSubject}
            onChange={handleEmailSubject}
          />
        </EuiFormRow>
        <EuiSpacer />
        <EuiFormRow
          label="Email body"
          labelAppend={<InsertPlaceholderPopover />}
        >
          <EuiTextArea
            fullWidth={true}
            placeholder="Body content"
            value={emailBody}
            onChange={handleEmailBody}
          />
        </EuiFormRow>
        <EuiSpacer size="xs" />
        <EuiCheckbox
          id="includeReportAsAttachment"
          label="Include report as attachment"
          checked={includeReportAsAttachment}
          onChange={handleIncludeReportAsAttachment}
        />
      </div>
    );
  };

  const emailDelivery = emailCheckbox ? <EmailDelivery /> : null;

  createReportDefinitionRequest['delivery'] = delivery;

  return (
    <EuiPageContent panelPaddingSize={'l'}>
      <EuiPageHeader>
        <EuiTitle>
          <h2>Delivery settings</h2>
        </EuiTitle>
      </EuiPageHeader>
      <EuiHorizontalRule />
      <EuiPageContentBody>
        <EuiFormRow label="Kibana recipients" helpText="Select or add users">
          <DeliveryRecipientsBox />
        </EuiFormRow>
        <EuiSpacer />
        <EuiCheckbox
          id="emailCheckboxDelivery"
          label="Email"
          checked={emailCheckbox}
          onChange={handleEmailCheckbox}
        />
        <EuiSpacer />
        {emailDelivery}
      </EuiPageContentBody>
    </EuiPageContent>
  );
}
