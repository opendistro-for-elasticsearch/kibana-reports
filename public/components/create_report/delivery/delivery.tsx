/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import {
  EuiFieldText,
  EuiFormRow,
  EuiPageHeader,
  EuiTitle,
  EuiPageContent,
  EuiPageContentBody,
  EuiHorizontalRule,
  EuiTextArea,
  EuiCheckbox,
  EuiSpacer,
  EuiComboBox,
  EuiText,
  EuiLink,
  EuiPopover,
  EuiListGroup,
} from '@elastic/eui';
import { DeliveryRecipientsBox } from './delivery_recipients_box';
import { email_recipient_options } from './delivery_constants';

const insert_placeholder_options = [
  {
    label: 'Report details URL',
    href: '#',
    iconType: 'link',
    size: 's',
  },
  {
    label: 'Report source URL',
    href: '#',
    iconType: 'link',
    size: 's',
  },
  {
    label: 'File download URL',
    href: '#',
    iconType: 'link',
    size: 's',
  },
  {
    label: 'Report creation timestamp',
    href: '#',
    iconType: 'clock',
    size: 's',
  },
];

export function ReportDelivery() {
  const [emailCheckbox, setEmailCheckbox] = useState(false);
  const [emailRecipients, setEmailRecipients] = useState();
  const [emailSubject, setEmailSubject] = useState('');
  const [emailBody, setEmailBody] = useState('');
  const [includeReportAsAttachment, setIncludeReportAsAttachment] = useState(
    false
  );
  const [insertPlaceholder, setInsertPlaceholder] = useState(false);

  const onChangeEmailCheckbox = (e) => {
    setEmailCheckbox(e.target.checked);
  };

  const onChangeEmailRecipients = (e) => {
    setEmailRecipients(e);
  };

  const onChangeEmailSubject = (e) => {
    setEmailSubject(e.target.value);
  };

  const onChangeEmailBody = (e) => {
    setEmailBody(e.target.value);
  };

  const onChangeIncludeReportAsAttachment = (e) => {
    setIncludeReportAsAttachment(e.target.checked);
  };

  const onInsertPlaceholderClick = () =>
    setInsertPlaceholder((insertPlaceholder) => !insertPlaceholder);
  const closeInsertPlaceholder = () => setInsertPlaceholder(false);

  const onCreateEmailRecipient = (searchValue, flattenedOptions = []) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
    };

    // Create the option if it doesn't exist.
    if (
      flattenedOptions.findIndex(
        (option) => option.label.trim().toLowerCase() === normalizedSearchValue
      ) === -1
    ) {
      email_recipient_options.push(newOption);
    }

    // Select the option.
    setEmailRecipients([...emailRecipients, newOption]);
  };

  const placeholderInsert = (
    <EuiText size="xs">
      <EuiLink onClick={onInsertPlaceholderClick}>Insert placeholder</EuiLink>
    </EuiText>
  );

  const InsertPlaceholderPopover = () => {
    return (
      <div>
        <EuiPopover
          button={placeholderInsert}
          isOpen={insertPlaceholder}
          closePopover={closeInsertPlaceholder}
        >
          <EuiListGroup listItems={insert_placeholder_options} />
        </EuiPopover>
      </div>
    );
  };

  const EmailDelivery = () => {
    return (
      <div>
        <EuiFormRow
          label="Email recipients"
          helpText="Select or add recipients"
        >
          <EuiComboBox
            placeholder={'add users here'}
            options={email_recipient_options}
            selectedOptions={emailRecipients}
            onChange={onChangeEmailRecipients}
            onCreateOption={onCreateEmailRecipient}
          />
        </EuiFormRow>
        <EuiSpacer />
        <EuiFormRow label="Email subject">
          <EuiFieldText
            placeholder="Subject line"
            value={emailSubject}
            onChange={onChangeEmailSubject}
          />
        </EuiFormRow>
        <EuiSpacer />
        <EuiFormRow
          label="Email body"
          labelAppend={<InsertPlaceholderPopover />}
        >
          <EuiTextArea
            fullWidth={true}
            placeholder="Body content"
            value={emailBody}
            onChange={onChangeEmailBody}
          />
        </EuiFormRow>
        <EuiSpacer size="xs" />
        <EuiCheckbox
          id="includeReportAsAttachment"
          label="Include report as attachment"
          checked={includeReportAsAttachment}
          onChange={onChangeIncludeReportAsAttachment}
        />
      </div>
    );
  };

  const email_delivery = emailCheckbox ? <EmailDelivery /> : null;

  return (
    <EuiPageContent panelPaddingSize={'l'}>
      <EuiPageHeader>
        <EuiTitle>
          <h2>Delivery settings</h2>
        </EuiTitle>
      </EuiPageHeader>
      <EuiHorizontalRule />
      <EuiPageContentBody>
        <EuiFormRow label="Kibana recipients" helpText="Select or add users">
          <DeliveryRecipientsBox />
        </EuiFormRow>
        <EuiSpacer />
        <EuiCheckbox
          id="emailCheckboxDelivery"
          label="Email"
          checked={emailCheckbox}
          onChange={onChangeEmailCheckbox}
        />
        <EuiSpacer />
        {email_delivery}
      </EuiPageContentBody>
    </EuiPageContent>
  );
}
