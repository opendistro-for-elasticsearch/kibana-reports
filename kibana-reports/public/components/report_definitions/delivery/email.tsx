/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiComboBox, EuiFieldText, EuiFormRow, EuiSpacer } from '@elastic/eui';
import React, { useEffect } from 'react';
import { useState } from 'react';
import ReactMDE from 'react-mde';
import { ReportDeliveryProps } from './delivery';
import { ChannelSchemaType, DeliverySchemaType } from 'server/model';
import { converter } from '../utils';
import { DELIVERY_TYPE_OPTIONS } from './delivery_constants';

export const EmailDelivery = (props: ReportDeliveryProps) => {
  const {
    edit,
    editDefinitionId,
    reportDefinitionRequest,
    httpClientProps,
    showEmailRecipientsError,
    emailRecipientsErrorMessage,
  } = props;

  const [emailRecipients, setEmailRecipients] = useState([]);
  const [selectedEmailRecipients, setSelectedEmailRecipients] = useState([]);
  const [emailSubject, setEmailSubject] = useState('');
  const [emailBody, setEmailBody] = useState('');
  const [selectedTab, setSelectedTab] = React.useState<'write' | 'preview'>(
    'write'
  );

  const handleCreateEmailRecipient = (
    searchValue: string,
    flattenedOptions = []
  ) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
    };

    // Create the option if it doesn't exist.
    if (
      flattenedOptions.findIndex(
        (option) => option.label.trim().toLowerCase() === normalizedSearchValue
      ) === -1
    ) {
      setEmailRecipients([...emailRecipients, newOption]);
    }

    handleSelectEmailRecipients([...selectedEmailRecipients, newOption]);
  };

  const handleSelectEmailRecipients = (e) => {
    setSelectedEmailRecipients(e);
    reportDefinitionRequest.delivery.delivery_params.recipients = e.map(
      (option) => option.label
    );
  };

  const handleEmailSubject = (e) => {
    setEmailSubject(e.target.value);
    reportDefinitionRequest.delivery.delivery_params.title = e.target.value;
  };

  const handleEmailBody = (e) => {
    setEmailBody(e);
    reportDefinitionRequest.delivery.delivery_params.textDescription = e;
    reportDefinitionRequest.delivery.delivery_params.htmlDescription = converter.makeHtml(
      e
    );
  };

  // TODO: need better handling when we add full support for kibana user report delivery
  const optionalMessageLabel = `Add optional message (${selectedTab} mode)`;

  const defaultEditDeliveryParams = (delivery: DeliverySchemaType) => {
    //TODO: need better handle?
    // if the original notification setting is kibana user
    if (delivery.delivery_type === DELIVERY_TYPE_OPTIONS[0].id) {
      defaultCreateDeliveryParams();
      delete reportDefinitionRequest.delivery.delivery_params.kibana_recipients;
    } else {
      //@ts-ignore
      const emailParams: ChannelSchemaType = delivery.delivery_params;
      const { recipients, title, textDescription } = emailParams;

      const recipientsOptions = recipients.map((email) => ({ label: email }));
      handleSelectEmailRecipients(recipientsOptions);
      setEmailRecipients(recipientsOptions);

      setEmailSubject(title);
      reportDefinitionRequest.delivery.delivery_params.title = title;
      handleEmailBody(textDescription);
    }
  };

  const defaultCreateDeliveryParams = () => {
    reportDefinitionRequest.delivery.delivery_params = {
      recipients: selectedEmailRecipients.map((option) => option.label),
      title: emailSubject,
      textDescription: emailBody,
      htmlDescription: converter.makeHtml(emailBody),
    };
  };

  useEffect(() => {
    if (edit) {
      httpClientProps
        .get(`../api/reporting/reportDefinitions/${editDefinitionId}`)
        .then(async (response) => {
          defaultEditDeliveryParams(response.report_definition.delivery);
        });
    } else {
      defaultCreateDeliveryParams();
    }
  }, []);

  return (
    <div>
      <EuiFormRow
        label="Email recipients"
        helpText="Select or add users"
        isInvalid={showEmailRecipientsError}
        error={emailRecipientsErrorMessage}
      >
        <EuiComboBox
          placeholder={'Add users here'}
          options={emailRecipients}
          selectedOptions={selectedEmailRecipients}
          onChange={handleSelectEmailRecipients}
          onCreateOption={handleCreateEmailRecipient}
          isClearable={true}
          data-test-subj="demoComboBox"
        />
      </EuiFormRow>
      <EuiSpacer />
      <EuiFormRow label="Email subject">
        <EuiFieldText
          placeholder="Subject line"
          value={emailSubject}
          onChange={handleEmailSubject}
        />
      </EuiFormRow>
      <EuiSpacer />
      <EuiFormRow label={optionalMessageLabel} fullWidth={true}>
        <ReactMDE
          value={emailBody}
          onChange={handleEmailBody}
          selectedTab={selectedTab}
          onTabChange={setSelectedTab}
          toolbarCommands={[
            ['header', 'bold', 'italic', 'strikethrough'],
            ['unordered-list', 'ordered-list', 'checked-list'],
          ]}
          generateMarkdownPreview={(markdown) =>
            Promise.resolve(converter.makeHtml(markdown))
          }
        />
      </EuiFormRow>
      <EuiSpacer size="xs" />
    </div>
  );
};
