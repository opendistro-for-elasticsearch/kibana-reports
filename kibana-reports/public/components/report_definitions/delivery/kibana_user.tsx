/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiComboBox, EuiFormRow } from '@elastic/eui';
import { useEffect } from 'react';
import { ReportDeliveryProps } from './delivery';

const KibanaUserDelivery = (props: ReportDeliveryProps) => {
  const {
    edit,
    editDefinitionId,
    reportDefinitionRequest,
    httpClientProps,
  } = props;

  const options = [];
  const [selectedOptions, setSelected] = useState([]);

  const onChangeDeliveryRecipients = (selectedOptions) => {
    setSelected(selectedOptions);
    reportDefinitionRequest.delivery.delivery_params.kibana_recipients = selectedOptions.map(
      (option) => option.label
    );
  };

  const onCreateDeliveryRecipientOption = (
    searchValue: string,
    flattenedOptions = []
  ) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
    };

    // Create the option if it doesn't exist.
    if (
      flattenedOptions.findIndex(
        (option) => option.label.trim().toLowerCase() === normalizedSearchValue
      ) === -1
    ) {
      options.push(newOption);
    }

    // Select the option.
    onChangeDeliveryRecipients([...selectedOptions, newOption]);
  };

  const defaultEditDeliveryParams = (deliveryParams) => {
    if (deliveryParams.kibana_recipients) {
      reportDefinitionRequest.delivery.delivery_params = {
        kibana_recipients: deliveryParams.kibana_recipients.map(
          (kibanaRecipient) =>
            onCreateDeliveryRecipientOption(kibanaRecipient, selectedOptions)
        ),
      };
    } else {
      const kibanaUserParams = {
        kibana_recipients: selectedOptions,
      };
      reportDefinitionRequest.delivery.delivery_params = kibanaUserParams;
    }
  };

  useEffect(() => {
    if (edit) {
      httpClientProps
        .get(`../api/reporting/reportDefinitions/${editDefinitionId}`)
        .then(async (response) => {
          defaultEditDeliveryParams(
            response.report_definition.delivery.delivery_params
          );
        });
    } else {
      reportDefinitionRequest.delivery.delivery_params = {
        kibana_recipients: selectedOptions,
      };
    }
  }, []);

  return (
    <EuiFormRow label="Kibana recipients" helpText="Select or add users">
      <EuiComboBox
        placeholder="Select or create options"
        options={options}
        selectedOptions={selectedOptions}
        onChange={onChangeDeliveryRecipients}
        onCreateOption={onCreateDeliveryRecipientOption}
        isClearable={true}
        data-test-subj="demoComboBox"
      />
    </EuiFormRow>
  );
};

export { KibanaUserDelivery };
