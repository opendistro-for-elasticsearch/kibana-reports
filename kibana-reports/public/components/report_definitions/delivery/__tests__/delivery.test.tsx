/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { ReportDelivery } from '../delivery';
import httpClientMock from '../../../../../test/httpMockClient';
import { act } from 'react-dom/test-utils';

const emptyRequest = {
  report_params: {
    report_name: '',
    report_source: '',
    description: '',
    core_params: {
      base_url: '',
      report_format: '',
      time_duration: '',
    },
  },
  delivery: {
    delivery_type: '',
    delivery_params: {},
  },
  trigger: {
    trigger_type: '',
    trigger_params: {},
  },
  time_created: 0,
  last_updated: 0,
  status: '',
};

const timeRange = {
  timeFrom: new Date(1234567800),
  timeTo: new Date(1234567890),
};

describe('<ReportDelivery /> panel', () => {
  test('render create component', () => {
    const { container } = render(
      <ReportDelivery
        edit={false}
        reportDefinitionRequest={emptyRequest}
        httpClientProps={httpClientMock}
        timeRange={timeRange}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('render edit component', async () => {
    const promise = Promise.resolve();
    let editReportDefinitionRequest = {
      report_params: {
        report_name: 'edit cron schedule component',
        report_source: 'Dashboard',
        description: '',
        core_params: {
          base_url: '',
          report_format: '',
          header: '',
          footer: '',
          time_duration: '',
        },
      },
      delivery: {
        delivery_type: '',
        delivery_params: {},
      },
      trigger: {
        trigger_type: 'On demand',
      },
    };

    httpClientMock.get = jest.fn().mockResolvedValue({
      report_definition: editReportDefinitionRequest,
    });

    const { container } = render(
      <ReportDelivery
        edit={true}
        editDefinitionId={'1'}
        reportDefinitionRequest={emptyRequest}
        httpClientProps={httpClientMock}
        timeRange={timeRange}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
    await act(() => promise);
  });
});
