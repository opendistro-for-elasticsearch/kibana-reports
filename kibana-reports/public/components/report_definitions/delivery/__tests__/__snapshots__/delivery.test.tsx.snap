// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportDelivery /> panel render create component 1`] = `
<div
  class="euiPanel euiPanel--paddingLarge euiPageContent"
  hidden=""
>
  <div
    class="euiPageHeader euiPageHeader--responsive"
  >
    <h2
      class="euiTitle euiTitle--medium"
    >
      Notification settings
    </h2>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginLarge"
  />
  <div
    class="euiPageContentBody"
  >
    <div
      class="euiCheckbox"
    >
      <input
        class="euiCheckbox__input"
        id="emailCheckboxDelivery"
        type="checkbox"
      />
      <div
        class="euiCheckbox__square"
      />
      <label
        class="euiCheckbox__label"
        for="emailCheckboxDelivery"
      >
        Add email recipients
      </label>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
  </div>
</div>
`;

exports[`<ReportDelivery /> panel render edit component 1`] = `
<div
  class="euiPanel euiPanel--paddingLarge euiPageContent"
  hidden=""
>
  <div
    class="euiPageHeader euiPageHeader--responsive"
  >
    <h2
      class="euiTitle euiTitle--medium"
    >
      Notification settings
    </h2>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginLarge"
  />
  <div
    class="euiPageContentBody"
  >
    <div
      class="euiCheckbox"
    >
      <input
        class="euiCheckbox__input"
        id="emailCheckboxDelivery"
        type="checkbox"
      />
      <div
        class="euiCheckbox__square"
      />
      <label
        class="euiCheckbox__label"
        for="emailCheckboxDelivery"
      >
        Add email recipients
      </label>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
  </div>
</div>
`;
